### PR TITLE
Add logic to convert Elastic Transcoder DAR to MediaConvert PAR.

### PR DESCRIPTION
**Pull request by wangchj**
*Issue #, if available:*

*Description of changes:*

Add logic to convert Elastic Transcoder DAR to MediaConvert PAR.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/get-par.js [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+ * Converts resolution and display aspect ratio (DAR) to pixel aspect ratio (PAR).
+ *
+ * @param {number} w Video resolution width.
+ * @param {number} h Video resolution height.
+ * @param {number} dw DAR width.
+ * @param {number} dh DAR height.
+ * @return {object} An object with the format {parNumerator, parDenominator} or undefined.
+ */
+function darToPar(w, h, dw, dh) {
+  const par = ((dw / dh) / (w / h));
```
Comment (bxthomas) [inline src/get-par.js]: We don't have to constrain ourselves to the enumeration in the MediaConvert UI.   How about something like this?

function darToPar(w, h, dw, dh) {
  const parNumerator = BigInt(dw) * BigInt(h);
  const parDenominator = BigInt(dh) * BigInt(w);

  // Use built-in Math.gcd (modern environments)
  const divisor = Math.gcd(parNumerator, parDenominator);

  return {
    parNumerator: parNumerator / divisor,
    parDenominator: parDenominator / divisor
  };
}

File: test/get-par.test.js [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    expect(res.parDenominator).toBe(1);
+  });
+});
+
+describe('darToPar()', () => {
+  it('should return PAR', () => {
+    const tests = [
+      {w: 200,  h: 200, dw: 2,  dh: 1, par: {parNumerator: 2, parDenominator: 1}},
+      {w: 1280, h: 720, dw: 16, dh: 9, par: {parNumerator: 1, parDenominator: 1}},
+      {w: 1,    h: 3,   dw: 5,  dh: 7, par: undefined},
```
Comment (bxthomas) [inline test/get-par.test.js]: Why can't we generate a valid par from this third test?

File: src/get-par.js (hunk 1 above)
Comment (wangchj) [reply src/get-par.js]: Yes I think using GCD is much cleaner than using a hardcoded map. I'll update this logic.

Also JavaScript has a large integer range. Do you think BigInt is required? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_SAFE_INTEGER

File: test/get-par.test.js (hunk 2 above)
Comment (wangchj) [reply test/get-par.test.js]: Same as comment above

File: src/get-par.js (hunk 1 above)
Comment (wangchj) [reply src/get-par.js]: Double checking, Elastic Transcoder does have validation on width and height:

> The value of the video width was '8000'. The video width is outside of the permitted range [128,4096].; The value of the video height was '8000'. The video height is outside of the permitted range [96,3072]. (Service: AmazonElasticTranscoder; Status Code: 400; Error Code: ValidationException; Request ID: ab508ac7-ee2b-4224-b10a-24e15dec215b; Proxy: null)

File: test/get-par.test.js [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+      aspectRatio: '4:3',
+      paddingPolicy: 'Pad',
+      _path: []
+    };
+
+    res = getPar(videoParams);
+
+    expect(res.parNumerator).toBe(4);
+    expect(res.parDenominator).toBe(3);
+  });
```
Comment (PavanRangudu) [inline test/get-par.test.js:51]: Can we add a test to verify that `getPar()` returns `undefined` when the required parameters for both methods are not provided, causing both methods to return `undefined`?

File: test/get-par.test.js [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+      paddingPolicy: 'Pad',
+      _path: []
+    };
+
+    let res = getPar(videoParams);
+
+    expect(res.parNumerator).toBe(1);
+    expect(res.parDenominator).toBe(1);
+
+    videoParams = {
```
Comment (PavanRangudu) [inline test/get-par.test.js:39]: Nit: this test can be splitted into a different `it()` block

File: src/get-par.js (hunk 1 above)
Comment (wangchj) [reply src/get-par.js]: Updated

File: test/get-par.test.js (hunk 2 above)
Comment (wangchj) [reply test/get-par.test.js]: Updated

File: test/get-par.test.js (hunk 3 above)
Comment (wangchj) [reply test/get-par.test.js:51]: `getPar()` has been updated to not to return undefined

File: test/get-par.test.js (hunk 4 above)
Comment (wangchj) [reply test/get-par.test.js:39]: I think they are parts of the same test

File: test/get-par.test.js [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    expect(res.parDenominator).toBe(1);
+  });
+});
+
+describe('darToPar()', () => {
+  it('should return PAR', () => {
+    const tests = [
+      {w: 200,  h: 200, dw: 2,  dh: 1, par: {parNumerator: 2,  parDenominator: 1}},
+      {w: 1280, h: 720, dw: 16, dh: 9, par: {parNumerator: 1,  parDenominator: 1}},
+      {w: 1,    h: 3,   dw: 5,  dh: 7, par: {parNumerator: 15, parDenominator: 7}},
```
Comment (PavanRangudu) [inline test/get-par.test.js:147]: @wangchj since you are going to add validations to width & height fields, we should probably add tests for 'zero', 'negative' and 'non-integer' inputs and make sure we are handling them gracefully.

File: test/get-par.test.js (hunk 3 above)
Comment (PavanRangudu) [reply test/get-par.test.js:43]: I think `getParFromAspectRatio` will still return `undefined` in below cases right?  

if (
    typeof videoParams.aspectRatio !== 'string' ||
    !videoParams.aspectRatio.match(aspectRatioRegex)
  ) {
    return;
  }


if (isNaN(resolution?.width) || isNaN(resolution?.height)) {
    addWarnMessage(
      [...videoParams._path, 'aspectRatio'],
      'Video aspect ratio cannot be converted to pixel aspect ratio without resolution. ' +
      'Video `aspectRatio` setting is ignored.'
    );

    return;
  }

Comment (wangchj) [reply test/get-par.test.js:51]: I have added a test for `undefined` case for `getPar()`.

The tests for `getParFromDisplayAspectRatio` and `getParFromAspectRatio` covers invalid input cases. I tried to avoid tests that test other functions.

File: test/get-par.test.js (hunk 5 above)
Comment (wangchj) [reply test/get-par.test.js:147]: I think the main use case for this script is to translate existing Elastic Transcoder  settings, which has already been validated by the API. For example, 

> The value of the video width was '8000'. The video width is outside of the permitted range [128,4096].; The value of the video height was '8000'. The video height is outside of the permitted range [96,3072]. (Service: AmazonElasticTranscoder; Status Code: 400; Error Code: ValidationException; Request ID: ab508ac7-ee2b-4224-b10a-24e15dec215b; Proxy: null)

Comment (wangchj) [reply test/get-par.test.js:147]: Added checks for invalid cases:

```
diff --git a/src/get-par.js b/src/get-par.js
index 4c7d564..f594340 100644
--- a/src/get-par.js
+++ b/src/get-par.js
@@ -51,7 +51,21 @@ function getParFromAspectRatio(videoParams) {
     return;
   }
 
-  return darToPar(resolution.width, resolution.height, parts[0], parts[1]);
+  const res = darToPar(resolution.width, resolution.height, parts[0], parts[1]);
+
+  if (!isFinite(res.parNumerator) || !isFinite(res.parDenominator) || res.parNumerator < 1 ||
+    res.parDenominator < 1)
+  {
+    addWarnMessage(
+      [...videoParams._path, 'aspectRatio'],
+      'Video aspect ratio cannot be converted to pixel aspect ratio. ' +
+      'Video `aspectRatio` setting is ignored.'
+    );
+
+    return;
+  }
+
+  return res;
 }
 
 /**
@@ -85,7 +99,21 @@ function getParFromDisplayAspectRatio(videoParams) {
     return;
   }
 
-  return darToPar(maxWidth, maxHeight, parts[0], parts[1]);
+  const res = darToPar(maxWidth, maxHeight, parts[0], parts[1]);
+
+  if (!isFinite(res.parNumerator) || !isFinite(res.parDenominator) || res.parNumerator < 1 ||
+  res.parDenominator < 1)
+  {
+    addWarnMessage(
+      [...videoParams._path, 'displayAspectRatio'],
+      'Video aspect ratio cannot be converted to pixel aspect ratio. ' +
+      'Video `displayAspectRatio` setting is ignored.'
+    );
+
+    return;
+  }
+
+  return res;
 }
 
 /**
diff --git a/test/get-par.test.js b/test/get-par.test.js
index b2eb138..3f729b5 100644
--- a/test/get-par.test.js
+++ b/test/get-par.test.js
@@ -53,23 +53,29 @@ describe('getPar()', () => {
 
 describe('getParFromAspectRatio()', () => {
   it('should return undefined when aspectRatio is auto or not defined', () => {
-    const tests = [undefined, 'auto', 123];
+    const tests = [undefined, 'auto', '0:0', '-4:-3', 123];
 
     for (const test of tests) {
-      const videoParams = {aspectRatio: test};
+      const videoParams = {aspectRatio: test, _path: []};
       expect(getParFromAspectRatio(videoParams)).toBeUndefined();
     }
   });
 
   it('should return undefined when resolution is not defined', () => {
-    const videoParams = {
-      aspectRatio: '16:9',
-      maxWidth: '1280',
-      maxHeight: '720',
-      _path: []
-    };
+    // Video resolutions
+    const tests = [undefined, '', '0x0', '-200x-200'];
 
-    expect(getParFromAspectRatio(videoParams)).toBeUndefined();
+    for (const test of tests) {
+      const videoParams = {
+        aspectRatio: '16:9',
+        maxWidth: '1280',
+        maxHeight: '720',
+        resolution: test,
+        _path: []
+      };
+
+      expect(getParFromAspectRatio(videoParams)).toBeUndefined();
+    }
   });
 
   it('should return PAR', () => {
@@ -87,28 +93,35 @@ describe('getParFromAspectRatio()', () => {
 
 describe('getParFromDisplayAspectRatio()', () => {
   it('should return undefined when displayAspectRatio is auto or not defined', () => {
-    const tests = [undefined, 'auto', 123];
+    const tests = [undefined, 'auto', '0:0', '-4:-3', 123];
 
     for (const test of tests) {
-      const videoParams = {displayAspectRatio: test};
+      const videoParams = {displayAspectRatio: test, _path: []};
       expect(getParFromDisplayAspectRatio(videoParams)).toBeUndefined();
     }
   });
 
   it('should return undefined when maxWith or maxHeight is not defined', () => {
-    let videoParams = {
-      displayAspectRatio: '16:9',
-      maxWidth: '1280',
-      _path: []
-    };
+    let tests = [
+      {maxWidth: undefined, maxHeight: undefined},
+      {maxWidth: '1280',    maxHeight: undefined},
+      {maxWidth: undefined, maxHeight: '720'},
+      {maxWidth: '0',       maxHeight: '0'},
+      {maxWidth: '-400',    maxHeight: '200'},
+      {maxWidth: '400',     maxHeight: '-200'},
+      {maxWidth: '-400',    maxHeight: '-200'},
+    ];
 
-    expect(getParFromDisplayAspectRatio(videoParams)).toBeUndefined();
+    for (const test of tests) {
+      let videoParams = {
+        displayAspectRatio: '16:9',
+        maxWidth: test.maxWidth,
+        maxHeight: test.maxHeight,
+        _path: []
+      };
 
-    videoParams = {
-      displayAspectRatio: '16:9',
-      maxHeight: '720',
-      _path: []
-    };
+      expect(getParFromDisplayAspectRatio(videoParams)).toBeUndefined();
+    }
   });
 
   it('should return PAR', () => {

```

Comment (PavanRangudu) [reply test/get-par.test.js:147]: LGTM. Thanks for adding these tests.

**Review by PavanRangudu — APPROVED** [review]
